ReactWindows/ReactNative.Shared/ReactInstanceManager.cs

```diff
@@ -328,50 +363,63 @@ public void OnResume(Action onBackPressed)
         /// associated with the provided root view will be started
         /// asynchronously. This view will then be tracked by this manager and
         /// in case of React instance restart, it will be re-attached.
+        /// WARNING! Has to be called by the thread associated with the view.
         /// </summary>
         /// <param name="rootView">The root view.</param>
-        public void AttachMeasuredRootView(ReactRootView rootView)
+        public async Task AttachMeasuredRootViewAsync(ReactRootView rootView)
         {
             if (rootView == null)
                 throw new ArgumentNullException(nameof(rootView));
 
-            DispatcherHelpers.AssertOnDispatcher();
+            DispatcherHelpers.AssertOnDispatcher(rootView);
             rootView.Children.Clear();
             rootView.ClearData();
-            _attachedRootViews.Add(rootView);
 
-            // If the React context is being created in the background, the
-            // JavaScript application will be started automatically when
-            // creation completes, as root view is part of the attached root
-            // view list.
-            var currentReactContext = _currentReactContext;
-            if (_contextInitializationTask == null && currentReactContext != null)
+            await DispatcherHelpers.CallOnDispatcher(() =>
             {
```

Comment:
I might be remembering incorrectly, but I remember getting exceptions when I tried to make a call from one dispatcher directly to another (something about the potential for deadlock). Is this no longer the case?

Reply:
Yeah, there are some ridiculous rules/behaviors.
1. You can access both <view>.Dispatcher and <view>.CoreWindow.Dispatcher from a worker thread.
2. You can access <view>.Dispatcher but you CAN'T <view>.CoreWindow from a different Dispatcher thread... You get that ASTA related exception with the deadlock scare in that case (I guess because the current Dispatcher is in a COM apartment already. I don't know the details of how .NET uses COM apartments, I know cross apartment calls are automatically marshalled through COM provided message loops so no deadlock should happen, maybe just reentrancy, but I digress).
I avoid using .CoreWindow for other reasons, like cases where that windo is not set yet, etc.

Other peculiarities:
- CoreApplication.MainView raises exception during Application constructor, yet CoreApplication.GetCurrentView() returns what seems to be the right thing (a view with IsMain == true)
- In some early scenarios that are about app being instantiated/activated from a different thread CoreApplication.MainView.Dispatcher fails with Exception(HRESULT_FROM(ELEMENT_NOT_FOUND)) :)
- The lack of SynchronizationContext in that case after init is done (we have that one bug/workaround for this).


---
In reply to: [165967906](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r165967906) [](ancestors = 165967906)